server/plugin.go

```diff
@@ -18,11 +35,237 @@ type Plugin struct {
+	msteamsAppClient      msteams.Client
+	msteamsBotClientMutex sync.Mutex
+	msteamsBotClient      msteams.Client
+
+	botID  string
+	userID string
+
+	subscriptionsToLinksMutex sync.Mutex
+	subscriptionsToLinks      map[string]ChannelLink
+	channelsLinked            map[string]ChannelLink
+
+	stopSubscriptions func()
 }
 
-// ServeHTTP demonstrates a plugin that handles HTTP requests by greeting the world.
 func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
-	fmt.Fprint(w, "Hello, world!")
+	api := NewAPI(p)
+	api.ServeHTTP(w, r)
+}
+
+func (p *Plugin) getURL() string {
+	config := p.API.GetConfig()
+	if strings.HasSuffix(*config.ServiceSettings.SiteURL, "/") {
+		return *config.ServiceSettings.SiteURL + "plugins/" + pluginID
+	}
+	return *config.ServiceSettings.SiteURL + "/plugins/" + pluginID
+}
+
+func (p *Plugin) connectTeamsAppClient() error {
+	p.msteamsAppClientMutex.Lock()
+	if p.msteamsAppClient == nil {
+		p.msteamsAppClient = msteams.NewApp(
+			p.configuration.TenantId,
+			p.configuration.ClientId,
+			p.configuration.ClientSecret,
+		)
+	}
+	err := p.msteamsAppClient.Connect()
+	if err != nil {
+		p.API.LogError("Unable to connect to the app client", "error", err)
+		return err
+	}
+	p.msteamsAppClientMutex.Unlock()
+	return nil
 }
 
-// See https://developers.mattermost.com/extend/plugins/server/reference/
+func (p *Plugin) connectTeamsBotClient() error {
+	p.msteamsBotClientMutex.Lock()
+	if p.msteamsBotClient == nil {
+		p.msteamsBotClient = msteams.NewBot(
+			p.configuration.TenantId,
+			p.configuration.ClientId,
+			p.configuration.ClientSecret,
+			p.configuration.BotUsername,
+			p.configuration.BotPassword,
+		)
+	}
+	err := p.msteamsBotClient.Connect()
+	if err != nil {
+		p.API.LogError("Unable to connect to the bot client", "error", err)
+		return err
+	}
+	p.msteamsBotClientMutex.Unlock()
+	return nil
+}
+
+func (p *Plugin) start() {
+	err := p.connectTeamsAppClient()
+	if err != nil {
+		p.API.LogError("Unable to connect to the msteams", "error", err)
+		return
+	}
+	err = p.connectTeamsBotClient()
+	if err != nil {
+		p.API.LogError("Unable to connect to the msteams", "error", err)
+		return
+	}
+
+	channelsLinkedData, appErr := p.API.KVGet("channelsLinked")
+	if appErr != nil {
+		p.API.LogError("Error getting the channels linked", "error", appErr)
+		return
+	}
+	channelsLinked := map[string]ChannelLink{}
+	err = json.Unmarshal(channelsLinkedData, &channelsLinked)
+	if err != nil {
+		p.API.LogError("Error getting the channels linked", "error", err)
+		return
+	}
+
+	p.channelsLinked = channelsLinked
+	p.subscriptionsToLinks = map[string]ChannelLink{}
+	ctx, stop := context.WithCancel(context.Background())
+	p.stopSubscriptions = stop
+	err = p.msteamsAppClient.ClearSubscriptions()
+	if err != nil {
+		p.API.LogError("Unable to clear all subscriptions", "error", err)
+	}
+	for _, link := range channelsLinked {
+		go p.subscribeToChannel(ctx, link)
+	}
```

Comment:
Just so I understand, why do we clear the subscriptions and re-instantiate them?

Reply:
I was trying to clean up everything on startup, probably there are better ways to do that. I'll give it an extra iteration to see how I can make it more robust and consistent (I'm not super happy with that either)

Reply:
Is this function safe in a HA environment?

Reply:
In fact, this is accidentally working in HA because you can only subscribe to 1 channel 1 time, so... even if you have multiple servers subscribing to it, it will work, and even if the clear on the other server remove that, the last one subscribing should work. Anyway, I should use a cluster mutex for this, to ensure that. I'm already having all the subscriptions abstracted in a different service so that is going to be easier now.